Upgrade to Ruby 3.2.2 and Rails 7

bruno opened the pull request:
This is an attempt to upgrade our fork to work with ruby 3.2.2 and rails 7 as part of our effort for our main application.

Some changes were backported from upstream 4261 and others are just about fixing warnings and complaints when running the unit tests.

Refactored the buildkite build process to use the matrix build and make it a bit more explicit.

Unit tests:

```
Finished in 7.239716232 seconds.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
4093 tests, 69078 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
565.35 tests/s, 9541.53 assertions/s
```

🎫 waysact/evergiving#6590

bruno committed:
Pick some upstream changes for rails and ruby

These changes allow us to run with Rails 7.x and Ruby 3.2.x. These changes come from upstream in PR 4261.

Refs waysact/evergiving#6590

bruno committed:
Bump up main dockerfile to our stable ruby

We’re still on 2.7.8 in our main app, even though this Dockerfile is not used in relation to our main app, we should keep updated.

Refs waysact/evergiving#6590

bruno committed:
A few additional responses that need to be allowed

These are not allowed by default.

Refs waysact/evergiving#6590

bruno committed:
Bump up Dockerfile to rails 7 and ruby 3.2.2

We only use this for local development, so this is handy.

Refs waysact/evergiving#6590

bruno committed:
Use the same base image as waysact

Using the `-slim` version was causing issues with OpenSSL

h/t @jscheid for the suggestion

Refs waysact/evergiving#6590

bruno committed:
Make extract_options available in BraintreeBlue

Refs waysact/evergiving#6590

bruno committed:
class_eval against the class and not the instance

Refs waysact/evergiving#6590

bruno committed:
Use the same base image as waysact

I was wondering why this was failing in CI and not when I ran the suite locally. 🙈

Refs waysact/evergiving#6590

bruno committed:
Add byebug gem to test goup

This is needed by remote_addon_payment_test.rb

Refs waysact/evergiving#6590

bruno committed:
Crediban requires a valid client_id for some calls

Crediban remote tests fail for certain operations if the ClientID supplied is not valid (empty). Ensure that we pass a correct value.

Note: the working credentials are not supplied in this test.

Refs waysact/evergiving#6590

bruno committed:
Bank options need to be merged in the right order

The `client_id_type: INT` is needed when interacting with ePayCo when testing banking (the API error says so) and the merge needs to be done in this order to preserve it over the one that already exists in `@options`

Refs waysact/evergiving#6590

bruno committed:
Re-write pipeline to use matrix build

Remove the shell script we used to test (which was just using `docker`) and replace it with a matrix build configuration in Buildkite to get more visibility over the combo of ruby and rails versions we test against.

Refs waysact/evergiving#6590

bruno committed:
Fix warnings for regex ambiguity

warning: ambiguity between regexp and two divisions: wrap regexp in parentheses or add a space after `/' operator

Refs waysact/evegiving#6590

bruno committed:
Remove unused var

Refs waysact/evergiving#6590

bruno committed:
Fix indentation of end and def keywords

Refs waysact/evergiving#6590

bruno committed:
message was never returned in this method

Refs waysact/evergiving#6590

bruno committed:
Fix another ambiguous regex warning

Refs waysact/evergiving#6590

bruno committed:
Fix rescue mismatched indentations

Refs waysact/evergiving#6590

bruno committed:
Remove unused url variable

The `url()` method is defined later in the file.

Refs waysact/evergiving#6590

bruno committed:
url was not returned properly in this method

Refs waysact/evergiving#6590

bruno committed:
Fix unused variables

Refs waysact/evergiving#6590

fbrusatti reviewed (approved):
lgtm

fbrusatti committed:
Remove invalid library requirement. Refs waysact/evergiving#6590

bruno committed:
Remove byebug from the Gemfile

Since ae8046c removed it, we no longer need to have it installed.

Refs waysact/evergiving#6590

jscheid commented on lib/active_merchant/billing/gateways/realex.rb:
```diff
@@ -472,6 +472,8 @@ def message_from(response)
           message = DECLINED
         end
+
+        message
```
It shouldn't matter because `x = y` returns `y`.

If this is for readability, a better improvement might be to simplify it to just:
```ruby
case response[:result]
when "00" then SUCCESS
when "101" then response[:message]
...
```

henare replied:
Nice catch, Julian. It looks like Bruno made this change to [suppress a warning](https://buildkite.com/waysact/active-merchant/builds/481#01894e35-f045-4140-bf24-c651df33d048/105-189). Your simplification makes sense to me and upstream has [made a similar change](https://github.com/activemerchant/active_merchant/commit/4abd2f0db48a6314ab16906525efb34f8b7e6bc5#diff-8f17a82e2a38e71c76ad882d3c1a6a684606e41ba57cf9d484365a3f9b14f523). Since Bruno is busy with the audit I've taken the liberty of [applying that change to this PR](https://github.com/waysact/active_merchant/pull/247/commits/73f97675f87efcdccab3f48a4261ad1e2d66d35c).

henare reviewed (approved):
Really nice rework of the Buildkite pipeline :clap:

henare committed:
Remove unnecessary variable assignment

This was also done upstream:

https://github.com/activemerchant/active_merchant/commit/4abd2f0db48a6314ab16906525efb34f8b7e6bc5#diff-8f17a82e2a38e71c76ad882d3c1a6a684606e41ba57cf9d484365a3f9b14f523

Refs waysact/evergiving#6590
